### PR TITLE
New plugin: gpu_nvml collects NVIDIA GPU stats.

### PR DESCRIPTION
**Pull request by qdbp**
This is a C plugin that efficiently collects various operational parameters of NVIDIA GPUs through direct invocation of the NVML API. The targeted use case is monitoring and debugging hardware status for long-running GPGPU jobs (e.g. a BOINC daemon).

Currently following stats are collected:

- fan speed
- core temperature and power consumption
- memory and streaming multiprocessor clocks
- percent GPU and memory loads

The NVML header is provided by CUDA, having which is a build requirement for this module.

### CODE REVIEW

**Comment by ccnifo** [comment]
Great work, thanks!
Maybe you should consider _nvmlDeviceGetMemoryErrorCounter_? We found this metric interesting to spot unexpected hardware errors ending in jobs being killed.

File: src/gpu_nvml.c [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+    int is_match = ((1 << ix) & conf_match_mask) || (conf_match_mask == 0);
+    if (conf_mask_is_exclude == !!is_match) {
+      continue;
+    }
+
+    nvmlDevice_t dev;
+    TRY(nvmlDeviceGetHandleByIndex(ix, &dev));
+
+    char dev_name[MAX_DEVNAME_LEN + 1];
```
Comment (octo) [inline src/gpu_nvml.c]: Initialize right here:

```c
char dev_name[MAX_DEVNAME_LEN + 1] = {0};
```

File: src/gpu_nvml.c [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    dev_name[0] = '\0';
+    TRY(nvmlDeviceGetName(dev, dev_name, MAX_DEVNAME_LEN));
```
Comment (octo) [inline src/gpu_nvml.c]: Please use `sizeof(dev_name)-1` instead.

File: src/gpu_nvml.c [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+    // Try to be as lenient as possible with the variety of devices that are
+    // out there, ignoring any NOT_SUPPORTED errors gently.
+    nvmlMemory_t meminfo;
+    TRYOPT(nvmlDeviceGetMemoryInfo(dev, &meminfo))
+    if (nv_status == NVML_SUCCESS) {
+      double pct_mem_used = 100. * (double)meminfo.used / meminfo.total;
+      nvml_submit(dev_name, "percent", "mem_used", WRAPGAUGE(pct_mem_used));
```
Comment (octo) [inline src/gpu_nvml.c]: Please consider to dispatch this as `memory-used` and `memory-free`, i.e. using the `memory` *type*, instead.

File: src/gpu_nvml.c [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+// that there are no more than 64 GPUs on this system.
+static uint64_t conf_match_mask = 0;
+static bool conf_mask_is_exclude = 0;
+
+static int nvml_config(const char *key, const char *value) {
+
+  unsigned long device_ix;
+  char *eptr;
+
+  if (strcasecmp(key, config_keys[0]) == 0) {
```
Comment (octo) [inline src/gpu_nvml.c]: I'd find this easier to read if you repeated the string literal `"GPUIndex"` here. Alternatively, add a comment to `config_keys` that its order matters – that is not normally the case.

File: src/gpu_nvml.c [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+  if ((nv_status = f) != NVML_SUCCESS &&                                       \
+      nv_status != NVML_ERROR_NOT_SUPPORTED) {                                 \
+    nv_errline = #f;                                                           \
+    goto catch;                                                                \
+  }
+
+#define TRY(f) TRY_CATCH(f, catch)
+#define TRYOPT(f) TRY_CATCH_OPTIONAL(f, catch)
+
+#define WRAPGAUGE(x) ((value_t){.gauge = (gauge_t)(x)})
```
Comment (octo) [inline src/gpu_nvml.c]: If you pass a `gauge_t` to `nvml_submit()` and do the conversion there, I think you can avoid this macro.

File: src/gpu_nvml.c [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+      return -1;
+    }
+    if (device_ix >= 64) {
+      ERROR("At most 64 GPUs (0 <= GPUIndex < 64) are supported!");
+      return -2;
+    }
+    conf_match_mask |= (1 << device_ix);
+  } else if (strcasecmp(key, config_keys[1])) {
+    if
+      IS_TRUE(value) { conf_mask_is_exclude = 1; }
```
Comment (octo) [inline src/gpu_nvml.c]:
```c
conf_mask_is_exclude = IS_TRUE(value);
```

File: src/gpu_nvml.c [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+static int nvml_config(const char *key, const char *value) {
+
+  unsigned long device_ix;
+  char *eptr;
+
+  if (strcasecmp(key, config_keys[0]) == 0) {
+    device_ix = strtoul(value, &eptr, 10);
+    if (eptr == value) {
+      ERROR("Failed to parse GPUIndex value \"%s\"", value);
```
Comment (octo) [inline src/gpu_nvml.c]: Please prefix error messages with the plugin name, e.g.:

```c
ERROR(PLUGIN_NAME ": Failed …")
```

File: configure.ac [hunk 8]
```diff
@@ -6735,6 +6788,7 @@ AC_PLUGIN([filecount],           [yes],                     [Count files in dire
 AC_PLUGIN([fscache],             [$plugin_fscache],         [fscache statistics])
 AC_PLUGIN([gmond],               [$with_libganglia],        [Ganglia plugin])
 AC_PLUGIN([gps],                 [$plugin_gps],             [GPS plugin])
+AC_PLUGIN([gpu_nvml],            [$with_cuda],              [NVIDIA GPU plugin])
```
Comment (octo) [inline configure.ac]: What is "nvml" standing for? Is this a well-known acronym / abbreviation? I'd find "gpu_nvidia" more descriptive, but I have next to no experience in GPU things ;)

**Pull request merged**